### PR TITLE
Typed find operations

### PR DESCRIPTION
**Pull request by haiodo**
Typed find operations. Now it is a bit more hard to shut into self foot.

Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

### CODE REVIEW

File: packages/core/src/__tests__/model.test.ts [hunk 1]
```diff
@@ -622,4 +622,16 @@ describe('mixin tools', () => {
...
+    console.log('cool', {
+      name: 's1',
+      users: { userId: 'qwe' }
+    } as DocumentQuery<Space>)
+
+    console.log('cool', {
+      name: 's1',
+      users: [{ userId: 'qwe' }]
+    } as DocumentQuery<Space>)
+  })
```
Comment (sumboid) [inline packages/core/src/__tests__/model.test.ts:636]: It seems to be not necessary :)

File: plugins/chunter/src/components/internal/ChatView.svelte [hunk 2]
```diff
@@ -37,9 +38,11 @@
   let messages: Message[] = []
 
   const userId = getUserId()
+  let ms: Promise<QueryUpdater<Message>>
 
-  $: ms = liveQuery(ms, chunter.class.Message, { _space: space._id }, (docs) => {
+  $: ms = liveQuery<Message>(ms, chunter.class.Message, { _space: space._id as Ref<Space> }, (docs) => {
     messages = docs
+    console.log(messages)
```
Comment (sumboid) [inline plugins/chunter/src/components/internal/ChatView.svelte:45]: Unnecessary log

File: packages/core/src/__tests__/model.test.ts [hunk 3]
```diff
@@ -60,27 +60,27 @@ describe('matching', () => {
   })
 
   it('apply string value', () => {
-    const clone = model.createDocument(taskIds.class.Task, doc1)
+    const clone = model.newDoc(taskIds.class.Task, doc1._id, doc1)
```
Comment (z2sx) [inline packages/core/src/__tests__/model.test.ts:63]: Why new method?

**Review by z2sx — CHANGES_REQUESTED** [review]

File: packages/core/src/__tests__/model.test.ts (hunk 3 above)
Comment (haiodo) [reply packages/core/src/__tests__/model.test.ts:63]: It was two methods doing same., I've removed one, question is naming only.

**Review by sumboid — CHANGES_REQUESTED** [review]

File: packages/core/src/__tests__/model.test.ts (hunk 3 above)
Comment (z2sx) [reply packages/core/src/__tests__/model.test.ts:63]: 1. They are different: `createDocument` has optional `id` generating new one when not provided, `newDoc` always requires `id`
2. If they are same, why new name, is `createDocument` bad?

Comment (z2sx) [reply packages/core/src/__tests__/model.test.ts:63]: I mean it seems `createDocument` used everywhere and has better signature, why removing it and replacing with `newDoc` everywhere? Or `newDoc` used more often?

File: plugins/chunter/src/components/internal/ChatView.svelte (hunk 2 above)
Comment (haiodo) [reply plugins/chunter/src/components/internal/ChatView.svelte]: Fixed, Thanks.

File: packages/core/src/__tests__/model.test.ts (hunk 1 above)
Comment (haiodo) [reply packages/core/src/__tests__/model.test.ts:638]: Fixed, Thanks.

**Review by sumboid — DISMISSED** [review]

File: packages/core/src/__tests__/model.test.ts (hunk 3 above)
Comment (haiodo) [reply packages/core/src/__tests__/model.test.ts:63]: I think, it should be named `create` to match .

```typescript
/**
 * Define operations with object modifications.
 */
export interface OperationProtocol {
  /**
   * Perform creation of new document. Object ID will be automatically generated and assigned to object.
   */
  create<T extends Doc> (_class: Ref<Class<T>>, values: AnyLayout | Doc): Promise<T>
```

Comment (haiodo) [reply packages/core/src/__tests__/model.test.ts]: Code is updated.

Comment (z2sx) [reply packages/core/src/__tests__/model.test.ts]: So idea is to rename well established `createDocument` method for reason to match something in `OperationProtocol`? How many times `OperationProtocol.create` used in the codebase? Why not rename `OperationProtocol.create` to `createDocument` if you want some names to match?

Comment (haiodo) [reply packages/core/src/__tests__/model.test.ts]: From other side, it seems semantic for this two methods are different. Will return as createDocument and think about name for OperationProtocol.create => store or something more clear.

File: docs/content/docs/advanced/operation-api.md [hunk 4]
```diff
@@ -70,7 +70,7 @@ cs.find(task.mixin.VersionedTask, { version: { version: "1.0.0" } })
 
 ```typescript
 export interface OperationProtocol {
-	create<T extends Doc> (_class: Ref<Class<T>>, values: AnyLayout | Doc): Promise<T>
+	create<T extends Doc> (_class: Ref<Class<T>>, values: Partial<Doc>): Promise<T>
```
Comment (z2sx) [inline docs/content/docs/advanced/operation-api.md:73]: Why `Partial<Doc>` and not `Partial<T>`? :) Also why `Partial<T>` and not `T`? :)

File: packages/core/src/model.ts [hunk 5]
```diff
@@ -198,13 +198,12 @@ export class Model implements Storage {
...
    * @param layout
    */
-  public newDoc<T extends Doc> (_class: Ref<Class<T>>, _id: Ref<Doc>, layout: AnyLayout): T {
-    const doc = (this.assign({}, _class, layout) as unknown) as T
-    doc._id = _id
-    return doc
+  public createDocument<T extends Doc> (_class: Ref<Class<T>>, layout: Partial<T>): T {
+    const doc = this.assign({}, _class, layout as unknown as AnyLayout)
+    doc._id = generateId()
+    return doc as unknown as T
```
Comment (z2sx) [inline packages/core/src/model.ts:206]: In the past `createDocument` was accepting optional `id` param. Is this not a case now?

File: packages/core/src/model.ts [hunk 6]
```diff
@@ -198,13 +198,12 @@ export class Model implements Storage {
...
    * Construct a new proper document with all desired fields.
    * @param _class
-   * @param _id
    * @param layout
    */
-  public newDoc<T extends Doc> (_class: Ref<Class<T>>, _id: Ref<Doc>, layout: AnyLayout): T {
-    const doc = (this.assign({}, _class, layout) as unknown) as T
-    doc._id = _id
-    return doc
+  public createDocument<T extends Doc> (_class: Ref<Class<T>>, layout: Partial<T>): T {
```
Comment (z2sx) [inline packages/core/src/model.ts:203]: Why layout is `Parial<T>` and not `Omit<T, keyof Doc>`?

File: packages/core/src/model.ts [hunk 7]
```diff
@@ -388,16 +390,6 @@ export class Model implements Storage {
...
-  createDocument<M extends Doc> (_class: Ref<Class<M>>, values: Omit<M, keyof Doc>, _id?: Ref<M>): M {
-    const layout = {
-      _class,
-      _id: _id ?? this.generateId()
-    }
-    this.assign(layout, _class, (values as unknown) as AnyLayout)
-    // this.add(layout)
-    return (layout as unknown) as M
-  }
-
```
Comment (z2sx) [inline packages/core/src/model.ts:400]: Why remove this implementation and create new same one above? What's wrong with this?

File: packages/domains/src/indices/title.ts [hunk 8]
```diff
@@ -59,7 +59,8 @@ export class TitleIndex implements DomainIndex {
   }
 
   async onCreate (ctx: TxContext, create: CreateTx): Promise<any> {
-    const obj = this.modelDb.newDoc(create._objectClass, create._objectId, create.object)
+    const obj = this.modelDb.createDocument(create._objectClass, create.object)
+    obj._id = create._objectId
```
Comment (z2sx) [inline packages/domains/src/indices/title.ts:63]: OK, here is the case, why remove optional `id` from `createDocument` and do this tricks?

File: packages/model/src/builder.ts [hunk 9]
```diff
@@ -100,8 +100,11 @@ class Builder {
     return { _class: core.class.Attribute, type } as unknown as Attribute
   }
 
-  createDocument<M extends Doc> (_class: Ref<Class<M>>, values: Omit<M, keyof Doc>, _id?: Ref<M>): M {
-    const doc = this.memdb.createDocument(_class, values, _id)
+  createDocument<M extends Doc> (_class: Ref<Class<M>>, values: Partial<M>, _id?: Ref<Doc>): M {
+    const doc = this.memdb.createDocument<M>(_class, values)
+    if (_id) {
+      doc._id = _id
+    }
```
Comment (z2sx) [inline packages/model/src/builder.ts:107]: Again, same: why remove optional id from `memdb.createDocument` and do this tricks?

**Review by z2sx — CHANGES_REQUESTED** [review]

File: docs/content/docs/advanced/operation-api.md (hunk 4 above)
Comment (haiodo) [reply docs/content/docs/advanced/operation-api.md:73]: Good catch, will fix

File: packages/core/src/model.ts (hunk 5 above)
Comment (haiodo) [reply packages/core/src/model.ts:206]: It is used only in model generation in few cases, will revert this change, thanks.

File: packages/core/src/model.ts (hunk 6 above)
Comment (haiodo) [reply packages/core/src/model.ts:203]: we need to specify _space, probable we should just omit _id, _class

File: packages/model/src/builder.ts (hunk 9 above)
Comment (haiodo) [reply packages/model/src/builder.ts:107]: Will revert this change.

File: packages/domains/src/indices/title.ts (hunk 8 above)
Comment (haiodo) [reply packages/domains/src/indices/title.ts:63]: Will revert this change.

File: packages/core/src/model.ts (hunk 7 above)
Comment (haiodo) [reply packages/core/src/model.ts:400]: It was just two of them mostly the same, just choose one on top of file :-)

File: docs/content/docs/advanced/operation-api.md (hunk 4 above)
Comment (z2sx) [reply docs/content/docs/advanced/operation-api.md:73]: That was a question actually :) `Partial<T>` or `Omit<T, keyof Doc>`?

File: packages/core/src/model.ts (hunk 7 above)
Comment (z2sx) [reply packages/core/src/model.ts:400]: Implementations are different, this one with optional `id`

File: docs/content/docs/advanced/operation-api.md (hunk 4 above)
Comment (haiodo) [reply docs/content/docs/advanced/operation-api.md:73]: Omit<T, keyof Doc> is good, but we need a _space to be specified. So I suppose we should add _space as parameter for create,update,etc and have Omit<T, keyof Doc>.

Comment (haiodo) [reply docs/content/docs/advanced/operation-api.md:73]: Introduced DocumentValue<T> to support embedded fields without _class specified.

File: packages/core/src/model.ts (hunk 5 above)
Comment (haiodo) [reply packages/core/src/model.ts]: reverted.

File: packages/core/src/model.ts (hunk 6 above)
Comment (haiodo) [reply packages/core/src/model.ts]: Introduce DocumentValue<T>

**Review by z2sx — APPROVED** [review]

**Pull request merged**